{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import glob\n",
    "import torch\n",
    "import matplotlib.pyplot as plt\n",
    "import random\n",
    "import pickle\n",
    "import matplotlib as mpl\n",
    "from mpl_toolkits.mplot3d import Axes3D\n",
    "import matplotlib.cm as cm\n",
    "from   sklearn.decomposition import PCA\n",
    "from sklearn.manifold import TSNE\n",
    "from   sklearn.manifold import TSNE, Isomap, LocallyLinearEmbedding, MDS, SpectralEmbedding\n",
    "from   sklearn.preprocessing import StandardScaler\n",
    "#!{sys.executable} -m pip install scikit-learn-extra\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_path = '/mnt/data/CAMI/DNABERT/output/'\n",
    "\n",
    "tensors = []\n",
    "contig_names = []\n",
    "for filename in glob.glob('/mnt/data/CAMI/DNABERT/output/*.pickle'):\n",
    "    with open(filename, 'rb') as f:\n",
    "        x = pickle.load(f)\n",
    "        tensors.extend(list(x.values()))\n",
    "        contig_names.extend(list(x.keys()))\n",
    "concat_tensors = torch.cat(tensors,0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "taxonomy = '/mnt/data/CAMI/data/short_read_oral/taxonomy.tsv'\n",
    "contig_to_genome = '/mnt/data/CAMI/data/short_read_oral/reformatted_manually_combined_gsa_mapping.tsv'\n",
    "\n",
    "contig_to_genome_df = pd.read_csv(contig_to_genome, sep='\\t', header=None)\n",
    "contig_to_genome_df = contig_to_genome_df.rename(columns={0: 'contig_name', 1: 'genome'})\n",
    "\n",
    "taxonomy_df = pd.read_csv(taxonomy, sep='\\t', header = None)\n",
    "taxonomy_df = taxonomy_df.rename(columns={0: 'genome', 1: 'species', 2: 'genus'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig_name</th>\n",
       "      <th>genome</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>species</th>\n",
       "      <th>genus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>S13C144379</td>\n",
       "      <td>OTU_97.10732.1</td>\n",
       "      <td>CP014228.1</td>\n",
       "      <td>848814</td>\n",
       "      <td>849755</td>\n",
       "      <td>Actinomyces radicidentis</td>\n",
       "      <td>Actinomyces</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>S13C4155</td>\n",
       "      <td>OTU_97.27895.0</td>\n",
       "      <td>CP010906.1</td>\n",
       "      <td>758589</td>\n",
       "      <td>760618</td>\n",
       "      <td>Campylobacter jejuni</td>\n",
       "      <td>Campylobacter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>S13C140957</td>\n",
       "      <td>OTU_97.37776.1</td>\n",
       "      <td>CP002094.1</td>\n",
       "      <td>2081039</td>\n",
       "      <td>2081268</td>\n",
       "      <td>Lactococcus lactis</td>\n",
       "      <td>Lactococcus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>S13C24289</td>\n",
       "      <td>OTU_97.34522.1</td>\n",
       "      <td>AM295007.1</td>\n",
       "      <td>1679584</td>\n",
       "      <td>1680153</td>\n",
       "      <td>Streptococcus pyogenes</td>\n",
       "      <td>Streptococcus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>S13C116303</td>\n",
       "      <td>OTU_97.23.0</td>\n",
       "      <td>LN831051.1</td>\n",
       "      <td>2043591</td>\n",
       "      <td>2043855</td>\n",
       "      <td>Streptococcus pneumoniae</td>\n",
       "      <td>Streptococcus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2964252</th>\n",
       "      <td>S8C75706</td>\n",
       "      <td>OTU_97.28883.0</td>\n",
       "      <td>CP000921.1</td>\n",
       "      <td>1377239</td>\n",
       "      <td>1379025</td>\n",
       "      <td>Streptococcus pneumoniae</td>\n",
       "      <td>Streptococcus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2964253</th>\n",
       "      <td>S8C46428</td>\n",
       "      <td>OTU_97.39134.0</td>\n",
       "      <td>FM204883.1</td>\n",
       "      <td>609148</td>\n",
       "      <td>609430</td>\n",
       "      <td>Streptococcus equi</td>\n",
       "      <td>Streptococcus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2964254</th>\n",
       "      <td>S8C149508</td>\n",
       "      <td>OTU_97.39522.0</td>\n",
       "      <td>CP012075.1</td>\n",
       "      <td>150371</td>\n",
       "      <td>152780</td>\n",
       "      <td>Prevotella fusca</td>\n",
       "      <td>Prevotella</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2964255</th>\n",
       "      <td>S8C112712</td>\n",
       "      <td>OTU_97.28231.0</td>\n",
       "      <td>HE858529.1</td>\n",
       "      <td>2154398</td>\n",
       "      <td>2155309</td>\n",
       "      <td>Streptococcus dysgalactiae</td>\n",
       "      <td>Streptococcus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2964256</th>\n",
       "      <td>S8C95887</td>\n",
       "      <td>OTU_97.7659.0</td>\n",
       "      <td>CP012071.1</td>\n",
       "      <td>56781</td>\n",
       "      <td>57260</td>\n",
       "      <td>Selenomonas sp. oral taxon 478</td>\n",
       "      <td>Selenomonas</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2964257 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        contig_name          genome           2        3        4  \\\n",
       "0        S13C144379  OTU_97.10732.1  CP014228.1   848814   849755   \n",
       "1          S13C4155  OTU_97.27895.0  CP010906.1   758589   760618   \n",
       "2        S13C140957  OTU_97.37776.1  CP002094.1  2081039  2081268   \n",
       "3         S13C24289  OTU_97.34522.1  AM295007.1  1679584  1680153   \n",
       "4        S13C116303     OTU_97.23.0  LN831051.1  2043591  2043855   \n",
       "...             ...             ...         ...      ...      ...   \n",
       "2964252    S8C75706  OTU_97.28883.0  CP000921.1  1377239  1379025   \n",
       "2964253    S8C46428  OTU_97.39134.0  FM204883.1   609148   609430   \n",
       "2964254   S8C149508  OTU_97.39522.0  CP012075.1   150371   152780   \n",
       "2964255   S8C112712  OTU_97.28231.0  HE858529.1  2154398  2155309   \n",
       "2964256    S8C95887   OTU_97.7659.0  CP012071.1    56781    57260   \n",
       "\n",
       "                                species          genus  \n",
       "0              Actinomyces radicidentis    Actinomyces  \n",
       "1                  Campylobacter jejuni  Campylobacter  \n",
       "2                    Lactococcus lactis    Lactococcus  \n",
       "3                Streptococcus pyogenes  Streptococcus  \n",
       "4              Streptococcus pneumoniae  Streptococcus  \n",
       "...                                 ...            ...  \n",
       "2964252        Streptococcus pneumoniae  Streptococcus  \n",
       "2964253              Streptococcus equi  Streptococcus  \n",
       "2964254                Prevotella fusca     Prevotella  \n",
       "2964255      Streptococcus dysgalactiae  Streptococcus  \n",
       "2964256  Selenomonas sp. oral taxon 478    Selenomonas  \n",
       "\n",
       "[2964257 rows x 7 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_df = pd.merge(contig_to_genome_df, taxonomy_df, how=\"left\", on=[\"genome\"])\n",
    "merged_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig_name</th>\n",
       "      <th>genome</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>species</th>\n",
       "      <th>genus</th>\n",
       "      <th>contig_lengths</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>S13C144379</td>\n",
       "      <td>OTU_97.10732.1</td>\n",
       "      <td>CP014228.1</td>\n",
       "      <td>848814</td>\n",
       "      <td>849755</td>\n",
       "      <td>Actinomyces radicidentis</td>\n",
       "      <td>Actinomyces</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>S13C4155</td>\n",
       "      <td>OTU_97.27895.0</td>\n",
       "      <td>CP010906.1</td>\n",
       "      <td>758589</td>\n",
       "      <td>760618</td>\n",
       "      <td>Campylobacter jejuni</td>\n",
       "      <td>Campylobacter</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>S13C140957</td>\n",
       "      <td>OTU_97.37776.1</td>\n",
       "      <td>CP002094.1</td>\n",
       "      <td>2081039</td>\n",
       "      <td>2081268</td>\n",
       "      <td>Lactococcus lactis</td>\n",
       "      <td>Lactococcus</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>S13C24289</td>\n",
       "      <td>OTU_97.34522.1</td>\n",
       "      <td>AM295007.1</td>\n",
       "      <td>1679584</td>\n",
       "      <td>1680153</td>\n",
       "      <td>Streptococcus pyogenes</td>\n",
       "      <td>Streptococcus</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>S13C116303</td>\n",
       "      <td>OTU_97.23.0</td>\n",
       "      <td>LN831051.1</td>\n",
       "      <td>2043591</td>\n",
       "      <td>2043855</td>\n",
       "      <td>Streptococcus pneumoniae</td>\n",
       "      <td>Streptococcus</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2964252</th>\n",
       "      <td>S8C75706</td>\n",
       "      <td>OTU_97.28883.0</td>\n",
       "      <td>CP000921.1</td>\n",
       "      <td>1377239</td>\n",
       "      <td>1379025</td>\n",
       "      <td>Streptococcus pneumoniae</td>\n",
       "      <td>Streptococcus</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2964253</th>\n",
       "      <td>S8C46428</td>\n",
       "      <td>OTU_97.39134.0</td>\n",
       "      <td>FM204883.1</td>\n",
       "      <td>609148</td>\n",
       "      <td>609430</td>\n",
       "      <td>Streptococcus equi</td>\n",
       "      <td>Streptococcus</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2964254</th>\n",
       "      <td>S8C149508</td>\n",
       "      <td>OTU_97.39522.0</td>\n",
       "      <td>CP012075.1</td>\n",
       "      <td>150371</td>\n",
       "      <td>152780</td>\n",
       "      <td>Prevotella fusca</td>\n",
       "      <td>Prevotella</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2964255</th>\n",
       "      <td>S8C112712</td>\n",
       "      <td>OTU_97.28231.0</td>\n",
       "      <td>HE858529.1</td>\n",
       "      <td>2154398</td>\n",
       "      <td>2155309</td>\n",
       "      <td>Streptococcus dysgalactiae</td>\n",
       "      <td>Streptococcus</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2964256</th>\n",
       "      <td>S8C95887</td>\n",
       "      <td>OTU_97.7659.0</td>\n",
       "      <td>CP012071.1</td>\n",
       "      <td>56781</td>\n",
       "      <td>57260</td>\n",
       "      <td>Selenomonas sp. oral taxon 478</td>\n",
       "      <td>Selenomonas</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2964257 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        contig_name          genome           2        3        4  \\\n",
       "0        S13C144379  OTU_97.10732.1  CP014228.1   848814   849755   \n",
       "1          S13C4155  OTU_97.27895.0  CP010906.1   758589   760618   \n",
       "2        S13C140957  OTU_97.37776.1  CP002094.1  2081039  2081268   \n",
       "3         S13C24289  OTU_97.34522.1  AM295007.1  1679584  1680153   \n",
       "4        S13C116303     OTU_97.23.0  LN831051.1  2043591  2043855   \n",
       "...             ...             ...         ...      ...      ...   \n",
       "2964252    S8C75706  OTU_97.28883.0  CP000921.1  1377239  1379025   \n",
       "2964253    S8C46428  OTU_97.39134.0  FM204883.1   609148   609430   \n",
       "2964254   S8C149508  OTU_97.39522.0  CP012075.1   150371   152780   \n",
       "2964255   S8C112712  OTU_97.28231.0  HE858529.1  2154398  2155309   \n",
       "2964256    S8C95887   OTU_97.7659.0  CP012071.1    56781    57260   \n",
       "\n",
       "                                species          genus  contig_lengths  \n",
       "0              Actinomyces radicidentis    Actinomyces             NaN  \n",
       "1                  Campylobacter jejuni  Campylobacter             NaN  \n",
       "2                    Lactococcus lactis    Lactococcus             NaN  \n",
       "3                Streptococcus pyogenes  Streptococcus             NaN  \n",
       "4              Streptococcus pneumoniae  Streptococcus             NaN  \n",
       "...                                 ...            ...             ...  \n",
       "2964252        Streptococcus pneumoniae  Streptococcus             NaN  \n",
       "2964253              Streptococcus equi  Streptococcus             NaN  \n",
       "2964254                Prevotella fusca     Prevotella             NaN  \n",
       "2964255      Streptococcus dysgalactiae  Streptococcus             NaN  \n",
       "2964256  Selenomonas sp. oral taxon 478    Selenomonas             NaN  \n",
       "\n",
       "[2964257 rows x 8 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "contig_lengths = \"/mnt/data/CAMI/DNABERT/contig_lengths.pickle\"\n",
    "with open(contig_lengths,'rb') as f:\n",
    "    contig_lengths = pickle.load(f)\n",
    "    \n",
    "contig_lengths = {k[1:]: v for k, v in contig_lengths.items()}\n",
    "merged_df[\"contig_lengths\"] = merged_df[\"contig_name\"].map(contig_lengths)\n",
    "merged_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig_name</th>\n",
       "      <th>genome</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>species</th>\n",
       "      <th>genus</th>\n",
       "      <th>contig_lengths</th>\n",
       "      <th>contig_idx</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1667473</th>\n",
       "      <td>S19C259815</td>\n",
       "      <td>OTU_97.41539.0</td>\n",
       "      <td>CP000260.1</td>\n",
       "      <td>1522037</td>\n",
       "      <td>1522883</td>\n",
       "      <td>Streptococcus pyogenes</td>\n",
       "      <td>Streptococcus</td>\n",
       "      <td>847.0</td>\n",
       "      <td>450415.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1667474</th>\n",
       "      <td>S19C345587</td>\n",
       "      <td>OTU_97.11620.0</td>\n",
       "      <td>CP016442.1</td>\n",
       "      <td>3161170</td>\n",
       "      <td>3161457</td>\n",
       "      <td>Burkholderia stabilis</td>\n",
       "      <td>Burkholderia</td>\n",
       "      <td>288.0</td>\n",
       "      <td>291286.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1667475</th>\n",
       "      <td>S19C625314</td>\n",
       "      <td>OTU_97.39134.1</td>\n",
       "      <td>CP003858.1</td>\n",
       "      <td>1362831</td>\n",
       "      <td>1363769</td>\n",
       "      <td>Streptococcus intermedius</td>\n",
       "      <td>Streptococcus</td>\n",
       "      <td>939.0</td>\n",
       "      <td>390114.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1667476</th>\n",
       "      <td>S19C433562</td>\n",
       "      <td>OTU_97.14178.0</td>\n",
       "      <td>CP003368.1</td>\n",
       "      <td>1390173</td>\n",
       "      <td>1390609</td>\n",
       "      <td>Prevotella dentalis</td>\n",
       "      <td>Prevotella</td>\n",
       "      <td>437.0</td>\n",
       "      <td>656162.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1667477</th>\n",
       "      <td>S19C496262</td>\n",
       "      <td>OTU_97.28043.0</td>\n",
       "      <td>CP016912.1</td>\n",
       "      <td>844138</td>\n",
       "      <td>844733</td>\n",
       "      <td>Burkholderia pseudomallei</td>\n",
       "      <td>Burkholderia</td>\n",
       "      <td>596.0</td>\n",
       "      <td>75861.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2347903</th>\n",
       "      <td>S19C71531</td>\n",
       "      <td>OTU_97.40690.0</td>\n",
       "      <td>CP009147.1</td>\n",
       "      <td>957418</td>\n",
       "      <td>979750</td>\n",
       "      <td>Burkholderia mallei</td>\n",
       "      <td>Burkholderia</td>\n",
       "      <td>22333.0</td>\n",
       "      <td>603531.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2347904</th>\n",
       "      <td>S19C197686</td>\n",
       "      <td>OTU_97.41539.0</td>\n",
       "      <td>CP000260.1</td>\n",
       "      <td>796679</td>\n",
       "      <td>797632</td>\n",
       "      <td>Streptococcus pyogenes</td>\n",
       "      <td>Streptococcus</td>\n",
       "      <td>954.0</td>\n",
       "      <td>240885.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2347905</th>\n",
       "      <td>S19C294810</td>\n",
       "      <td>OTU_97.29395.1</td>\n",
       "      <td>CP000381.1</td>\n",
       "      <td>1154419</td>\n",
       "      <td>1154669</td>\n",
       "      <td>Neisseria meningitidis</td>\n",
       "      <td>Neisseria</td>\n",
       "      <td>251.0</td>\n",
       "      <td>36709.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2347906</th>\n",
       "      <td>S19C452300</td>\n",
       "      <td>OTU_97.8703.1</td>\n",
       "      <td>CP017050.1</td>\n",
       "      <td>2539173</td>\n",
       "      <td>2539461</td>\n",
       "      <td>Burkholderia pseudomallei</td>\n",
       "      <td>Burkholderia</td>\n",
       "      <td>289.0</td>\n",
       "      <td>4199.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2347907</th>\n",
       "      <td>S19C209855</td>\n",
       "      <td>OTU_97.9324.1</td>\n",
       "      <td>CP008782.1</td>\n",
       "      <td>1001199</td>\n",
       "      <td>1001560</td>\n",
       "      <td>Burkholderia pseudomallei</td>\n",
       "      <td>Burkholderia</td>\n",
       "      <td>362.0</td>\n",
       "      <td>335655.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>680401 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        contig_name          genome           2        3        4  \\\n",
       "1667473  S19C259815  OTU_97.41539.0  CP000260.1  1522037  1522883   \n",
       "1667474  S19C345587  OTU_97.11620.0  CP016442.1  3161170  3161457   \n",
       "1667475  S19C625314  OTU_97.39134.1  CP003858.1  1362831  1363769   \n",
       "1667476  S19C433562  OTU_97.14178.0  CP003368.1  1390173  1390609   \n",
       "1667477  S19C496262  OTU_97.28043.0  CP016912.1   844138   844733   \n",
       "...             ...             ...         ...      ...      ...   \n",
       "2347903   S19C71531  OTU_97.40690.0  CP009147.1   957418   979750   \n",
       "2347904  S19C197686  OTU_97.41539.0  CP000260.1   796679   797632   \n",
       "2347905  S19C294810  OTU_97.29395.1  CP000381.1  1154419  1154669   \n",
       "2347906  S19C452300   OTU_97.8703.1  CP017050.1  2539173  2539461   \n",
       "2347907  S19C209855   OTU_97.9324.1  CP008782.1  1001199  1001560   \n",
       "\n",
       "                           species          genus  contig_lengths  contig_idx  \n",
       "1667473     Streptococcus pyogenes  Streptococcus           847.0    450415.0  \n",
       "1667474      Burkholderia stabilis   Burkholderia           288.0    291286.0  \n",
       "1667475  Streptococcus intermedius  Streptococcus           939.0    390114.0  \n",
       "1667476        Prevotella dentalis     Prevotella           437.0    656162.0  \n",
       "1667477  Burkholderia pseudomallei   Burkholderia           596.0     75861.0  \n",
       "...                            ...            ...             ...         ...  \n",
       "2347903        Burkholderia mallei   Burkholderia         22333.0    603531.0  \n",
       "2347904     Streptococcus pyogenes  Streptococcus           954.0    240885.0  \n",
       "2347905     Neisseria meningitidis      Neisseria           251.0     36709.0  \n",
       "2347906  Burkholderia pseudomallei   Burkholderia           289.0      4199.0  \n",
       "2347907  Burkholderia pseudomallei   Burkholderia           362.0    335655.0  \n",
       "\n",
       "[680401 rows x 9 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "contig_name_to_idx = {v[1:]: i for i, v in enumerate(contig_names)}\n",
    "merged_df[\"contig_idx\"] = merged_df[\"contig_name\"].apply(lambda x: contig_name_to_idx.get(x))\n",
    "non_aligned = ~merged_df[\"contig_idx\"].isnull()\n",
    "aligned_tensor_df = merged_df[non_aligned]\n",
    "aligned_tensor_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3018058.0\n"
     ]
    }
   ],
   "source": [
    "column = aligned_tensor_df['contig_lengths']\n",
    "max_value = column.max()\n",
    "print(max_value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig_name</th>\n",
       "      <th>genome</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>species</th>\n",
       "      <th>genus</th>\n",
       "      <th>contig_lengths</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>contig_idx</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>S19C531701</td>\n",
       "      <td>OTU_97.28043.1</td>\n",
       "      <td>CP012517.1</td>\n",
       "      <td>1058437</td>\n",
       "      <td>1058708</td>\n",
       "      <td>Burkholderia pseudomallei</td>\n",
       "      <td>Burkholderia</td>\n",
       "      <td>272.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>S19C531702</td>\n",
       "      <td>OTU_97.27554.0</td>\n",
       "      <td>FN665654.1</td>\n",
       "      <td>303528</td>\n",
       "      <td>304852</td>\n",
       "      <td>Clostridioides difficile</td>\n",
       "      <td>Clostridioides</td>\n",
       "      <td>1325.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>S19C531703</td>\n",
       "      <td>OTU_97.9324.1</td>\n",
       "      <td>CP008782.1</td>\n",
       "      <td>1755368</td>\n",
       "      <td>1755617</td>\n",
       "      <td>Burkholderia pseudomallei</td>\n",
       "      <td>Burkholderia</td>\n",
       "      <td>250.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>S19C531704</td>\n",
       "      <td>OTU_97.27207.0</td>\n",
       "      <td>AP013072.1</td>\n",
       "      <td>1019916</td>\n",
       "      <td>1020409</td>\n",
       "      <td>Streptococcus anginosus</td>\n",
       "      <td>Streptococcus</td>\n",
       "      <td>494.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>S19C531705</td>\n",
       "      <td>OTU_97.934.0</td>\n",
       "      <td>CP008695.1</td>\n",
       "      <td>345079</td>\n",
       "      <td>345321</td>\n",
       "      <td>Streptococcus pyogenes</td>\n",
       "      <td>Streptococcus</td>\n",
       "      <td>243.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>680396</th>\n",
       "      <td>S19C275396</td>\n",
       "      <td>OTU_97.43518.0</td>\n",
       "      <td>CP019469.1</td>\n",
       "      <td>2197272</td>\n",
       "      <td>2197996</td>\n",
       "      <td>Clostridioides difficile</td>\n",
       "      <td>Clostridioides</td>\n",
       "      <td>725.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>680397</th>\n",
       "      <td>S19C275397</td>\n",
       "      <td>OTU_97.13253.0</td>\n",
       "      <td>LT671674.1</td>\n",
       "      <td>1375069</td>\n",
       "      <td>1375372</td>\n",
       "      <td>Streptococcus suis</td>\n",
       "      <td>Streptococcus</td>\n",
       "      <td>304.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>680398</th>\n",
       "      <td>S19C275398</td>\n",
       "      <td>OTU_97.8361.0</td>\n",
       "      <td>CP003087.1</td>\n",
       "      <td>2947729</td>\n",
       "      <td>2948014</td>\n",
       "      <td>Burkholderia sp. YI23</td>\n",
       "      <td>Burkholderia</td>\n",
       "      <td>286.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>680399</th>\n",
       "      <td>S19C275399</td>\n",
       "      <td>OTU_97.42543.1</td>\n",
       "      <td>AP012053.1</td>\n",
       "      <td>2176805</td>\n",
       "      <td>2177415</td>\n",
       "      <td>Streptococcus gallolyticus</td>\n",
       "      <td>Streptococcus</td>\n",
       "      <td>611.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>680400</th>\n",
       "      <td>S19C275400</td>\n",
       "      <td>OTU_97.13576.0</td>\n",
       "      <td>CP002833.1</td>\n",
       "      <td>1568353</td>\n",
       "      <td>1568695</td>\n",
       "      <td>Burkholderia pseudomallei</td>\n",
       "      <td>Burkholderia</td>\n",
       "      <td>343.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>680401 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           contig_name          genome           2        3        4  \\\n",
       "contig_idx                                                             \n",
       "0           S19C531701  OTU_97.28043.1  CP012517.1  1058437  1058708   \n",
       "1           S19C531702  OTU_97.27554.0  FN665654.1   303528   304852   \n",
       "2           S19C531703   OTU_97.9324.1  CP008782.1  1755368  1755617   \n",
       "3           S19C531704  OTU_97.27207.0  AP013072.1  1019916  1020409   \n",
       "4           S19C531705    OTU_97.934.0  CP008695.1   345079   345321   \n",
       "...                ...             ...         ...      ...      ...   \n",
       "680396      S19C275396  OTU_97.43518.0  CP019469.1  2197272  2197996   \n",
       "680397      S19C275397  OTU_97.13253.0  LT671674.1  1375069  1375372   \n",
       "680398      S19C275398   OTU_97.8361.0  CP003087.1  2947729  2948014   \n",
       "680399      S19C275399  OTU_97.42543.1  AP012053.1  2176805  2177415   \n",
       "680400      S19C275400  OTU_97.13576.0  CP002833.1  1568353  1568695   \n",
       "\n",
       "                               species           genus  contig_lengths  \n",
       "contig_idx                                                              \n",
       "0            Burkholderia pseudomallei    Burkholderia           272.0  \n",
       "1             Clostridioides difficile  Clostridioides          1325.0  \n",
       "2            Burkholderia pseudomallei    Burkholderia           250.0  \n",
       "3              Streptococcus anginosus   Streptococcus           494.0  \n",
       "4               Streptococcus pyogenes   Streptococcus           243.0  \n",
       "...                                ...             ...             ...  \n",
       "680396        Clostridioides difficile  Clostridioides           725.0  \n",
       "680397              Streptococcus suis   Streptococcus           304.0  \n",
       "680398           Burkholderia sp. YI23    Burkholderia           286.0  \n",
       "680399      Streptococcus gallolyticus   Streptococcus           611.0  \n",
       "680400       Burkholderia pseudomallei    Burkholderia           343.0  \n",
       "\n",
       "[680401 rows x 8 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aligned_tensor_index_df = aligned_tensor_df.astype({\"contig_idx\": int})\n",
    "aligned_tensor_index_df = aligned_tensor_index_df.set_index(\"contig_idx\").sort_index()\n",
    "aligned_tensor_index_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "680401\n",
      "431137\n"
     ]
    }
   ],
   "source": [
    "# Filter all contigs > 512\n",
    "\n",
    "print(len(aligned_tensor_index_df))\n",
    "filtered_512_df = aligned_tensor_index_df[aligned_tensor_index_df[\"contig_lengths\"] < 512]\n",
    "print(len(filtered_512_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_kmeans():\n",
    "    all_concat_np = concat_tensors.detach().numpy()\n",
    "    \n",
    "    pca = PCA(n_components=2)\n",
    "    pca.fit(all_concat_np)\n",
    "    projection = pca.transform(all_concat_np)\n",
    "    \n",
    "    \n",
    "    from sklearn.cluster import KMeans\n",
    "    wcss = []\n",
    "    for i in range(100,700):\n",
    "        kmeans_pca = KMeans(n_clusters=i, init='k-means++', random_state=None)\n",
    "        kmeans_pca.fit(projection)\n",
    "        wcss.append(kmeans_pca.inertia_)\n",
    "\n",
    "    plt.plot(range(100,700), wcss, marker = 'o', linestyle = '--')\n",
    "    \n",
    "    #kmeans_pca = KMeans(n_clusters = 3, init = 'k-means++', random_state=None).fit(projection)\n",
    "    #print(kmeans_pca.cluster_centers_)\n",
    "    #print(kmeans_pca.inertia_)\n",
    "    \n",
    "    #from sklearn_extra.cluster import KMedoids\n",
    "    #ss = []\n",
    "    #for i in range(1,10):\n",
    "        #kmedoids_pca = KMedoids(n_clusters=2, random_state=0)\n",
    "        #kmedoids_pca.fit(projection)\n",
    "        #ss.append(kmedoids_pca.inertia_)\n",
    "    \n",
    "    #plt.plot(range(1,10), ss, marker = 'o', linestyle = '--')\n",
    "\n",
    "    #kmedoids_pca = KMedoids(n_clusters=2, random_state=0).fit(projection)\n",
    "    #print(kmedoids_pca.cluster_centers_)\n",
    "    #print(kmedoids_pca.inertia_)\n",
    "    \n",
    "plot_kmeans()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "MIN_SIZE_OF_GROUP = 100\n",
    "NUM_TO_GROUPS_TO_PLOT = 10\n",
    "GROUP_KEY = \"species\"\n",
    "\n",
    "genome_group_indices = []\n",
    "i = 0\n",
    "groups = list(filtered_512_df.groupby(GROUP_KEY))\n",
    "random.shuffle(groups)\n",
    "for x_name, x in groups:\n",
    "    if i >= 10:\n",
    "        break\n",
    "        \n",
    "    group_size = len(x)\n",
    "    if group_size < MIN_SIZE_OF_GROUP or group_size > 1000:\n",
    "        continue\n",
    "        \n",
    "    genome_group_indices.extend(x.index.tolist())\n",
    "    i += 1\n",
    "\n",
    "print(len(genome_group_indices))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_pca():\n",
    "    all_concat_np = concat_tensors.detach().numpy()\n",
    "    genome_group_np = all_concat_np[genome_group_indices]\n",
    "    \n",
    "    pca = PCA(n_components=2)\n",
    "    pca.fit(genome_group_np)\n",
    "    projection = pca.transform(genome_group_np)\n",
    "    \n",
    "    genome_to_color_id = {k: i for k, i in zip(filtered_512_df.loc[genome_group_indices][GROUP_KEY].unique(), range(10))}\n",
    "    targets = filtered_512_df.loc[genome_group_indices][GROUP_KEY].apply(lambda x: genome_to_color_id[x]).tolist()\n",
    "    labels = filtered_512_df.loc[genome_group_indices][GROUP_KEY].unique().tolist()\n",
    "    plt.figure(figsize=(10, 10))\n",
    "    scatter = plt.scatter(projection[:, 0], projection[:, 1], alpha=0.9, s=5.0, c=targets, cmap='tab10')\n",
    "    plt.legend(handles=scatter.legend_elements()[0], labels=labels)\n",
    "    \n",
    "plot_pca()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_kmeans():\n",
    "    all_concat_np = concat_tensors.detach().numpy()\n",
    "    genome_group_np = all_concat_np[genome_group_indices]\n",
    "    \n",
    "    pca = PCA(n_components=2)\n",
    "    pca.fit(genome_group_np)\n",
    "    projection = pca.transform(genome_group_np)\n",
    "    \n",
    "    \n",
    "    from sklearn.cluster import KMeans\n",
    "    wcss = []\n",
    "    for i in range(1,10):\n",
    "        kmeans_pca = KMeans(n_clusters=i, init='k-means++', random_state=None)\n",
    "        kmeans_pca.fit(projection)\n",
    "        wcss.append(kmeans_pca.inertia_)\n",
    "\n",
    "    plt.plot(range(1,10), wcss, marker = 'o', linestyle = '--')\n",
    "    \n",
    "    kmeans_pca = KMeans(n_clusters = 3, init = 'k-means++', random_state=None).fit(projection)\n",
    "    print(kmeans_pca.cluster_centers_)\n",
    "    print(kmeans_pca.inertia_)\n",
    "    \n",
    "    from sklearn_extra.cluster import KMedoids\n",
    "    ss = []\n",
    "    for i in range(1,10):\n",
    "        kmedoids_pca = KMedoids(n_clusters=2, random_state=0)\n",
    "        kmedoids_pca.fit(projection)\n",
    "        ss.append(kmedoids_pca.inertia_)\n",
    "    \n",
    "    plt.plot(range(1,10), ss, marker = 'o', linestyle = '--')\n",
    "\n",
    "    kmedoids_pca = KMedoids(n_clusters=2, random_state=0).fit(projection)\n",
    "    print(kmedoids_pca.cluster_centers_)\n",
    "    print(kmedoids_pca.inertia_)\n",
    "    \n",
    "plot_kmeans()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_tsne():    \n",
    "    all_concat_np = concat_tensors.detach().numpy()\n",
    "    genome_group_np = all_concat_np[genome_group_indices]\n",
    "    \n",
    "    #NUM_SUBSAMPLE = 10000\n",
    "    subsample_indices = list(range(len(genome_group_np)))\n",
    "    #random.shuffle(subsample_indices)\n",
    "    # subsample_indices =  subsample_indices[:NUM_SUBSAMPLE]\n",
    "    \n",
    "    subsampled_genome_groups_np = genome_group_np[subsample_indices]\n",
    "    tsne = TSNE(n_components=2, perplexity=30)\n",
    "    projection = tsne.fit_transform(subsampled_genome_groups_np)\n",
    "    \n",
    "    genome_to_color_id = {k: i for k, i in zip(filtered_512_df.loc[genome_group_indices].iloc[subsample_indices][GROUP_KEY].unique(), range(10))}\n",
    "    targets = filtered_512_df.loc[genome_group_indices].iloc[subsample_indices][GROUP_KEY].apply(lambda x: genome_to_color_id[x]).tolist()\n",
    "    labels = filtered_512_df.loc[genome_group_indices][GROUP_KEY].unique().tolist()\n",
    "    plt.figure(figsize=(10, 10))\n",
    "    scatter = plt.scatter(projection[:, 0], projection[:, 1], alpha=0.9, s=3.0, c=targets, cmap='tab10')\n",
    "    plt.legend(handles=scatter.legend_elements()[0], labels=labels)\n",
    "\n",
    "plot_tsne()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import vamb\n",
    "\n",
    "def create_contig_file_list(path_to_contig_file):\n",
    "    contig_list = []\n",
    "    with open(path_to_contig_file, 'r') as fp:\n",
    "        lines = fp.readlines()\n",
    "        for line in lines:\n",
    "            line = line.rstrip()\n",
    "            contig_list.append(line)\n",
    "    return contig_list\n",
    "\n",
    "mincontiglength=10\n",
    "contig_file_list='/mnt/data/CAMI/vamb/workflow/contigs.txt'\n",
    "file_list = create_contig_file_list(contig_file_list)\n",
    "\n",
    "tnfs_per_fasta = []\n",
    "contignames_per_fasta = []\n",
    "for fasta in file_list:\n",
    "    with vamb.vambtools.Reader(fasta, 'rb') as tnffile:\n",
    "        tnfs, contignames, contiglengths = vamb.parsecontigs.read_contigs(tnffile, minlength=mincontiglength)\n",
    "        tnfs_per_fasta.append(tnfs)\n",
    "        contignames_per_fasta.extend(contignames)\n",
    "\n",
    "tnfs_per_fasta = np.concatenate(tnfs_per_fasta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(contignames_per_fasta))\n",
    "print(len(tnfs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# contignames to be sorted like filtered_512_df\n",
    "\n",
    "contig_name_to_idx = {v: i for i, v in enumerate(contignames_per_fasta)}\n",
    "vamb_aligned_tensor_df[\"contig_vamb_idx\"] = aligned_tensor_df[\"contig_name\"].map(contig_name_to_idx)\n",
    "vamb_aligned_tensor_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "non_aligned = ~vamb_aligned_tensor_df[\"contig_vamb_idx\"].isnull()\n",
    "vam_aligned_reset_tensor_df = vamb_aligned_tensor_df[non_aligned]\n",
    "vamb_aligned_tensor_df = vamb_aligned_tensor_df.astype({\"contig_vamb_idx\": int})\n",
    "vamb_aligned_tensor_df = vamb_aligned_tensor_df.astype({\"contig_idx\": int})\n",
    "vamb_aligned_tensor_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vamb_aligned_tensor_df_2 = vamb_aligned_tensor_df.set_index(\"contig_idx\").sort_index()\n",
    "vamb_aligned_tensor_df_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(vamb_aligned_tensor_df))\n",
    "vamb_filtered_512_df = vamb_aligned_tensor_df_2[vamb_aligned_tensor_df_2[\"contig_lengths\"] < 512]\n",
    "\n",
    "print(len(vamb_filtered_512_df))\n",
    "vamb_filtered_512_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "MIN_SIZE_OF_GROUP = 100\n",
    "MAX_SIZE_OF_GROUP = 1000\n",
    "NUM_TO_GROUPS_TO_PLOT = 10\n",
    "GROUP_KEY = \"species\"\n",
    "\n",
    "genome_group_indices = []\n",
    "i = 0\n",
    "groups = list(vamb_filtered_512_df.groupby(GROUP_KEY))\n",
    "random.shuffle(groups)\n",
    "for x_name, x in groups:\n",
    "    if i >= 10:\n",
    "        break\n",
    "        \n",
    "    group_size = len(x)\n",
    "    if group_size < MIN_SIZE_OF_GROUP or group_size > MAX_SIZE_OF_GROUP:\n",
    "        continue\n",
    "        \n",
    "    genome_group_indices.extend(x.index.tolist())\n",
    "    i += 1\n",
    "\n",
    "print(len(genome_group_indices))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_pca_matrix(matrix):\n",
    "    pca = PCA(n_components=2)\n",
    "    pca.fit(matrix)\n",
    "    projection = pca.transform(matrix)\n",
    "    \n",
    "    genome_to_color_id = {k: i for k, i in zip(vamb_filtered_512_df.loc[genome_group_indices][GROUP_KEY].unique(), range(10))}\n",
    "    targets = vamb_filtered_512_df.loc[genome_group_indices][GROUP_KEY].apply(lambda x: genome_to_color_id[x]).tolist()\n",
    "    labels = vamb_filtered_512_df.loc[genome_group_indices][GROUP_KEY].unique().tolist()\n",
    "    plt.figure(figsize=(10, 10))\n",
    "    scatter = plt.scatter(projection[:, 0], projection[:, 1], alpha=0.9, s=5.0, c=targets, cmap='tab10')\n",
    "    plt.legend(handles=scatter.legend_elements()[0], labels=labels)\n",
    "\n",
    "tnf_matrix = tnfs_per_fasta[vamb_filtered_512_df.loc[genome_group_indices][\"contig_vamb_idx\"].tolist()]\n",
    "plot_pca_matrix(tnf_matrix)\n",
    "\n",
    "all_concat_np = concat_tensors.detach().numpy()\n",
    "genome_group_np = all_concat_np[genome_group_indices]\n",
    "plot_pca_matrix(genome_group_np)\n",
    "\n",
    "tnf_matrix = (tnf_matrix) / (tnf_matrix.std())\n",
    "genome_group_np = (genome_group_np) / (genome_group_np.std())\n",
    "\n",
    "cat_tnf_dna = np.concatenate([genome_group_np, tnf_matrix], axis=-1)\n",
    "plot_pca_matrix(cat_tnf_dna)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_tsne_matrix(matrix):    \n",
    "    tsne = TSNE(n_components=2, perplexity=30)\n",
    "    projection = tsne.fit_transform(matrix)\n",
    "    \n",
    "    genome_to_color_id = {k: i for k, i in zip(vamb_filtered_512_df.loc[genome_group_indices][GROUP_KEY].unique(), range(10))}\n",
    "    targets = vamb_filtered_512_df.loc[genome_group_indices][GROUP_KEY].apply(lambda x: genome_to_color_id[x]).tolist()\n",
    "    labels = vamb_filtered_512_df.loc[genome_group_indices][GROUP_KEY].unique().tolist()\n",
    "    plt.figure(figsize=(10, 10))\n",
    "    scatter = plt.scatter(projection[:, 0], projection[:, 1], alpha=0.9, s=3.0, c=targets, cmap='tab10')\n",
    "    plt.legend(handles=scatter.legend_elements()[0], labels=labels)\n",
    "\n",
    "tnf_matrix = tnfs_per_fasta[vamb_filtered_512_df.loc[genome_group_indices][\"contig_vamb_idx\"].tolist()]\n",
    "plot_tsne_matrix(tnf_matrix)\n",
    "\n",
    "all_concat_np = concat_tensors.detach().numpy()\n",
    "genome_group_np = all_concat_np[genome_group_indices]\n",
    "plot_tsne_matrix(genome_group_np)\n",
    "\n",
    "tnf_matrix = (tnf_matrix) / (tnf_matrix.std())\n",
    "genome_group_np = (genome_group_np) / (genome_group_np.std())\n",
    "cat_tnf_dna = np.concatenate([genome_group_np, tnf_matrix], axis=-1)\n",
    "plot_tsne_matrix(cat_tnf_dna)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#torch.load('/mnt/data/CAMI/DNABERT/pretrained_models/4-new-12w-0/pytorch_model.bin', map_location=torch.device('cpu'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import torch \n",
    "from transformers import BertTokenizer, BertforSequenceClassification"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
